src/libAtomVM/nifs.c

```diff
@@ -1285,10 +1285,23 @@ static term nif_erlang_send_2(Context *ctx, int argc, term argv[])
 {
     UNUSED(argc);
 
-    term pid_term = argv[0];
-    VALIDATE_VALUE(pid_term, term_is_pid);
+    term send_too = argv[0];
+    int local_process_id;
+    if (term_is_pid(send_too) == 1) {
+        local_process_id = term_to_local_process_id(send_too);
+    } else if (term_is_atom(send_too) == true) {
+        int atom_index = term_to_atom_index(send_too);
+        int local_pid = globalcontext_get_registered_process(ctx->global, atom_index);
+        bool is_active = globalcontext_process_exists(ctx->global, local_pid);
+        if (UNLIKELY(is_active != true)) {
+            RAISE_ERROR(BADARG_ATOM);
```

Comment:
There is a race condition here.

- the process can be registered with a different name before the message is sent
- the process can die before the message is sent.

My intuition is you need to hold the lock on the process table, maybe by creating a new globalcontext_get_registered_process variant that does that, similar to `globalcontext_get_process_lock`

The OTP behavior is as follows regarding registered names and dead processes:

```
1> Pid = spawn(fun() -> ok end). 
<0.81.0>
2> Pid ! hello.
hello
3> spawn(fun() -> erlang:register(foo, self()) end).
<0.84.0>
4> foo ! hello.
** exception error: bad argument
     in operator  !/2
        called as foo ! hello
```

Reply:
I thought we did not allow registering a process with a second name:
https://github.com/atomvm/AtomVM/blob/29825e4bf18cb98409680a5dcd7687f2fb055213/src/libAtomVM/nifs.c#LL906C1-L911C6

This would seem to imply that we also have race conditions in the following functions, (possibly more):

- `nif_erlang_register_2`
- `nif_erlang_whereis_1`
- `nif_erlang_is_process_alive_1`

Or am I misunderstanding the nature of the race here?

Reply:
We have other issues indeed, for example names are not unregistered when the process dies.

The race condition is the following: you test if the process is active and return BADARG if it is not, following the semantic of BEAM. But what happens if the process becomes inactive after line 1297? Your code silently succeeds.

Likewise, you get the pid of a process by name, but many things can happen after you got this information, you are not holding any lock. For example the name may be unregistered and another process can be registered with the same name before your message is sent.

globalcontext_send_message holds a lock, but the lock probably should be acquired earlier.

Reply:
Thank you for the clarification. That was what I thought about the race condition. I believe the functions I listed above all suffer from similar race conditions.

I was aware that processes are not unregistered when they die, I realized that when I added the ability to unregister a process, but I could not find a solution for unregistering processes when they die... so I left that for a separate PR, hopefully by someone with a better understanding of the inner workings of the BEAM.

Reply:
Unregistering when a process dies indeed belongs to a separate PR but I'm confident you could author that. It's a matter of adding code to
`context_destroy`.

Reply:
Thank you for the hint on unregistering processes when they die. I will investigate that!

I do have one more follow-up on the race condition... Is holding a lock on the process table enough here, or should I also hold a lock on the registered processes table as well?

Reply:
It's always better to have only one lock. You may modify (if needed) the registering code to make sure it locks (only) the process table.